File: api/ts/day3.ts
import { NowRequest, NowResponse } from "@vercel/node";
import { StarResult } from "./shared/data";

export default (request: NowRequest, response: NowResponse) => {
  let input = request.body && request.body.input;
  let responseMessage: StarResult;

  const solve = (): StarResult => {
    if (input) {
      const lines = input.split(",");

      const calc = function (right, down) {
        let x = 0;
        let y = 0;
        let trees = 0;

        while (y < lines.length - down) {
          x += right;
          y += down;
          if (lines[y][x % lines[y].length] == "#") trees++;
        }

        return trees;
      };

      let silver = calc(3, 1);

      let gold = 1;

      let xVals = [1, 3, 5, 7, 1];
      let yVals = [1, 1, 1, 1, 2];
      for (let i = 0; i < xVals.length; i++) {
        gold *= calc(xVals[i], yVals[i]);
      }

      return { silver: silver.toString(), gold: gold.toString() };
    } else return { silver: "No input", gold: "No input" };
  };

  responseMessage = solve();
  response.status(200).json(responseMessage);
};

File: api/py/day3.py
from http.server import BaseHTTPRequestHandler
from datetime import datetime
import json

def calc(lines, right: int, down: int) -> int:
    x = 0
    y = 0
    trees = 0

    while y < len(lines) - down: 
        x += right
        y += down
        if lines[y][x % len(lines[y])] == "#":
            trees += 1
    return trees

class handler(BaseHTTPRequestHandler):

  def do_POST(self):
    content_len = int(self.headers.get('Content-Length'))
    post_body = self.rfile.read(content_len)
    body = json.loads(post_body)
    data = body['input']
    if data:
        lines = data.split(";")
        silver = calc(lines, 3, 1)
        gold = 1

        xVals = [1, 3, 5, 7, 1]
        yVals = [1, 1, 1, 1, 2]
        for i in range(len(xVals)):
            gold *= calc(lines, xVals[i], yVals[i])

        result = { "silver": str(silver), "gold": str(gold) }
    else:
        result = { "silver": "No input", "gold": "No input" }
    
    self.send_response(200)
    self.send_header('Content-Type', 'application/json')
    self.end_headers()
    self.wfile.write(json.dumps(result).encode('utf_8'))
    return